private/buf/bufsync/syncer.go

```diff
@@ -0,0 +1,174 @@
+		s.knownTagsByCommitHash[commitHash.Hex()] = append(s.knownTagsByCommitHash[commitHash.Hex()], tag)
+		return nil
+	}); err != nil {
+		return fmt.Errorf("process tags: %w", err)
+	}
+	// TODO: sync other branches
+	for _, branch := range []string{s.repo.BaseBranch()} {
+		// TODO: resume from last sync point
+		if err := s.repo.ForEachCommit(branch, func(commit git.Commit) error {
+			for _, module := range s.modulesToSync {
+				if err := s.visitCommit(
+					ctx,
+					module,
+					branch,
+					commit,
+					pushFunc,
+				); err != nil {
+					return fmt.Errorf("process commit %s: %w", commit.Hash().Hex(), err)
+				}
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("process commits: %w", err)
+		}
+	}
+	return nil
+}
+
+// visitCommit looks for the module in the commit, and if found tries to validate it.
+// If it is valid, it invokes `pushFunc`.
+//
+// It does not return errors on invalid modules, but it will return any errors from
+// `pushFunc` as those may be transient.
+func (s *syncer) visitCommit(
+	ctx context.Context,
+	module syncableModule,
+	branch string,
+	commit git.Commit,
+	pushFunc PushFunc,
+) error {
+	sourceBucket, err := s.storageGitProvider.NewReadBucket(
+		commit.Tree(),
+		storagegit.ReadBucketWithSymlinksIfSupported(),
+	)
+	if err != nil {
+		return err
+	}
+	sourceBucket = storage.MapReadBucket(sourceBucket, storage.MapOnPrefix(module.dir))
+	foundModule, err := bufconfig.ExistingConfigFilePath(ctx, sourceBucket)
+	if err != nil {
+		return err
+	}
+	if foundModule == "" {
+		// We did not find a module. Carry on to the next commit.
+		s.logger.Debug(
+			"module not found, skipping commit",
+			zap.String("commit", commit.Hash().String()),
+			zap.String("module", module.dir),
+		)
+		return nil
+	}
+	sourceConfig, err := bufconfig.GetConfigForBucket(ctx, sourceBucket)
+	if err != nil {
+		// We found a module but the module config is invalid. We can warn on this
+		// and carry on. Note that because of resumption, we will typically only come
+		// across this commit once, we will not log this warning again.
+		s.logger.Warn(
+			"invalid module",
+			zap.String("commit", commit.Hash().String()),
+			zap.String("module", module.dir),
+			zap.Error(err),
+		)
+		return nil
+	}
+	if sourceConfig.ModuleIdentity == nil {
+		// Unnamed module. Carry on.
+		s.logger.Debug(
+			"unnamed module, skipping commit",
+			zap.String("commit", commit.Hash().String()),
+			zap.String("module", module.dir),
+		)
+		return nil
+	}
+	moduleIdentity := sourceConfig.ModuleIdentity
+	if module.identityOverride != nil {
+		moduleIdentity = module.identityOverride
+	}
+	builtModule, err := bufmodulebuild.BuildForBucket(
+		ctx,
+		sourceBucket,
+		sourceConfig.Build,
+	)
+	if err != nil {
+		// We failed to build the module. We can warn on this
+		// and carry on. Note that because of resumption, we will typically only come
+		// across this commit once, we will not log this warning again.
+		s.logger.Warn(
```

Comment:
This is something that is conceivably part of control flow - it is conceivable that the caller of `bufsync` will want to know that a module was not built, and take some type of action. By logging here, we've effectively swallowed the warning. Instead, such data should be returned up the stack in a manner that allows the caller to decide what to do (including potentially log it)